src/DependencyInjection/DI/test/ServiceProviderContainerTests.cs

```diff
@@ -346,6 +346,52 @@ public void ProviderScopeDisposeThrowsWhenOnlyDisposeAsyncImplemented()
                 exception.Message);
         }
 
+        [Fact]
+        public void SingletonServiceCreatedFromFactoryIsDisposedWhenContainerIsDisposed()
+        {
+            // Arrange
+            var collection = new TestServiceCollection();
+            collection.AddSingleton(_ => new TestDisposable());
+            var provider = this.CreateServiceProvider(collection);
+
+            // Act
+            var service = provider.GetService<TestDisposable>();
+            ((IDisposable)provider).Dispose();
+
+            // Assert
+            Assert.True(service.IsDisposed);
+        }
+
+        [Fact]
+        public void SingletonServiceCreatedFromInstanceIsNotDisposedWhenContainerIsDisposed()
+        {
+            // Arrange
+            var collection = new TestServiceCollection();
+            collection.AddSingleton(new TestDisposable());
+            var provider = this.CreateServiceProvider(collection);
+
+            // Act
+            var service = provider.GetService<TestDisposable>();
+            ((IDisposable)provider).Dispose();
+
+            // Assert
+            Assert.False(service.IsDisposed);
+        }
+
+        internal class TestDisposable : IDisposable
+        {
+            public bool IsDisposed { get; private set; }
+
+            public void Dispose()
+            {
+                this.IsDisposed = true;
+            }
+        }
+
+        internal class TestServiceCollection : List<ServiceDescriptor>, IServiceCollection
```

Comment:
What's this for?

Reply:
It was meant to be an implementation of `IServiceCollection` (inheriting from `List<ServiceDescriptor>` is an easy way of implementing `IServiceCollection` for this purpose) that we could use for our tests, but this comment has made me realize that there is a default implementation of `IServiceCollection` called [`ServiceCollection`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.servicecollection?view=aspnetcore-3.0) that can be used for exactly this purpose.  I feel silly now.

Thanks for pointing this out, too!

Unrelated, but also in https://github.com/tomakita/Extensions/commit/0672203deab9b0ac7ef9660b0ab1f4bd88dfe958: I've changed the accessibility of `TestDisposable` to private, since it doesn't need to be internal.